### PR TITLE
Implementation of the dual slope in the clusterizer for Phase2 IT 

### PR DESCRIPTION
**Pull request by avartak**
The dual slope feature in the phase2 inner tracker digitizer has been implemented in the clusterizer code. Studies of this implementation have been reported in several presentation within the Tracker DPG. 

https://indico.cern.ch/event/688846/
https://indico.cern.ch/event/688869/

Note that a corresponding PR for the implementation of the dual slope in the digitizer code has also been submitted : 
https://github.com/cms-sw/cmssw/pull/24325

These are to be taken together. 

People closely following these developments are : 
@suchandradutta , @emiglior , @boudoul

### CODE REVIEW

File: RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc [hunk 1]
```diff
@@ -55,6 +55,10 @@ PixelThresholdClusterizer::PixelThresholdClusterizer
     theStackADC_( conf.exists("AdcFullScaleStack") ? conf.getParameter<int>("AdcFullScaleStack") : 255 ),
     theFirstStack_( conf.exists("FirstStackLayer") ? conf.getParameter<int>("FirstStackLayer") : 5 ),
     theElectronPerADCGain_( conf.exists("ElectronPerADCGain") ? conf.getParameter<double>("ElectronPerADCGain") : 135. ),
+    doPhase2Calibration( conf.exists("Phase2Calibration") ? conf.getParameter<bool>("Phase2Calibration") : false),
```
Comment (perrotta) [inline RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: Please add all these parameters to the fillDescription() method, and remove the check whether they exist.
Please do the same also for the ones above ("AdcFullScaleStack", "FirstStackLayer", "ElectronPerADCGain") even if not introduced by this PR

**Comment by slava77** [comment]
assign upgrade

File: RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc [hunk 2]
```diff
@@ -232,33 +236,45 @@ void PixelThresholdClusterizer::copy_to_buffer( DigiIterator begin, DigiIterator
...
+    } else {
+      int i=0;
+      for(DigiIterator di = begin; di != end; ++di) {
+        auto adc = di->adc();
+        const float gain = theElectronPerADCGain_; // default: 1 ADC = 135 electrons
+        const float pedestal = 0.; //
+        electron[i] = int(adc * gain + pedestal);
+        if (layer_>=theFirstStack_) {
+        if (theStackADC_==1&&adc==1) {
+          electron[i] = int(255*135); // Arbitrarily use overflow value.
```
Comment (kpedro88) [inline RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: make named constants for magic numbers 255 and 135

File: RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc [hunk 3]
```diff
@@ -314,6 +330,28 @@ int PixelThresholdClusterizer::calibrate(int adc, int col, int row)
...
 
+  if (doPhase2Calibration) {
+
+    const float gain = theElectronPerADCGain_;
+    int p2rm = (phase2ReadoutMode < -1 ? -1 : phase2ReadoutMode);
+    if (p2rm > 10) p2rm = 10;
+    const int dualslopeparam = abs(p2rm);
+    const int dualslope = int(dualslopeparam <= 1 ? 1. : pow(2, dualslopeparam-1));
+
+    if (adc < phase2KinkADC) electrons = int(adc * gain - gain * 0.5);
```
Comment (kpedro88) [inline RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: could simplify to `(adc - 0.5)*gain` (and similarly a few lines down)

File: RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc [hunk 4]
```diff
@@ -232,33 +236,45 @@ void PixelThresholdClusterizer::copy_to_buffer( DigiIterator begin, DigiIterator
...
+      if (layer_==1) {
+        (*theSiPixelGainCalibrationService_).calibrate(detid_,begin,end,theConversionFactor_L1, theOffset_L1,electron);
+      } else {
+        (*theSiPixelGainCalibrationService_).calibrate(detid_,begin,end,theConversionFactor,    theOffset,  electron);
+      }
+    } else {
+      int i=0;
+      for(DigiIterator di = begin; di != end; ++di) {
+        auto adc = di->adc();
+        const float gain = theElectronPerADCGain_; // default: 1 ADC = 135 electrons
```
Comment (perrotta) [inline RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: This and the following assignment can be moved outside the for loop

**Comment by perrotta** [comment]
ping @avartak

**Comment by avartak** [comment]
Hi Andrea,

I have implemented almost all the comments. There are one or two things I want to discuss with @emiglior , @boudoul before pushing them to git.

**Comment by perrotta** [comment]
@avartak , @emiglior , @boudoul : any news?

**Comment by emiglior** [comment]
@perrotta we are investigating the (unexpected) differences observed in PR #24325

**Comment by emiglior** [comment]
@perrotta @avartak  @boudoul 
Comparison with baseline in PR #24325 is now OK

**Comment by perrotta** [comment]
@emiglior @avartak : I imagine that "now" in "Comparison with baseline in PR #24325 is now OK" means "as soon as the relevant commits are merged in this pull request".
Please merge those commits, that should also fulfill the other comments received during the review, so that we can evaluate them.

**Comment by emiglior** [comment]
@perrotta PR #24325 concerns only the digitizer while this PR concerns only the clusterizer.
In both cases the default config is supposed to reproduce the TDR/current CMSSW behaviour (a part a small fix in the config file of the digitizer)
Shouldn't be possible to merge them independently?

**Comment by perrotta** [comment]
Ah, ok Ernesto, my fault: I imagined you were talking about this PR, not another one!
Then, for this one, we just wait for the review comments being implemented before proceding.

**Comment by avartak** [comment]
The PR has been updated addressing the comments received.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
Hi Adish.

I see two kinds of issues to be fixed.

First of all, there are two instances of siPixelClusters defined in CMSSW, and this can confuse the parser:
- the one created automatically by the fillDescription() method of PixelThresholdClusterizer (which is an instance of PixelThresholdClusterizer)
- the one defined explicitely in RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py (which is an instance of SiPixelClusterProducer)

The parameters of the PixelThresholdClusterizer are copied in RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py, and *at the very least* you should repeat the very same list there. I am convinced, however, that this duplication could be avoided by making siPixelClusters importing directly the default configuration of the PixelThresholdClusterizer, maybe in the same way as it is done now for the SiPixelGainCalibrationServiceParameters.

Honestly I don't know by heart how one can do it in our framework. @Dr15Jones do you have an idea how to import the same parameters defined in https://github.com/cms-sw/cmssw/pull/24329/files#diff-b27bb6c187ebafb0818f16f4bb0932d8R94 (i.e., defined by fillDescription() and not organized as a PSet) into RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py?

Once solved it, the HLT parser will eventually be able to correctly parse the configuration whenever the new code will be available in the release. In the meanwhile, you need to customize the present HLT configuration with the newly added parameters. This can be done in the file
```
HLTrigger/Configuration/python/customizeHLTforCMSSW.py
```
(please see the examples therein for doing it)
@Martin-Grunewald

**Comment by perrotta** [comment]
@avartak 

After discussing the issue with Slava (thanks), I think we can proceed simply as follow:
- Add the missing parameters to RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py
- Add a dedicated customization for the missing parameters in HLT to HLTrigger/Configuration/python/customizeHLTforCMSSW.py

There is no reason, nor advantage, for doing anything more complicated.

In addition, it would be good to avoid giving the same name to different configurations, even if not explicitly used. Therefore, you can think about modifying to something different from "siPixelClusters" in https://github.com/cms-sw/cmssw/pull/24329/files#diff-b27bb6c187ebafb0818f16f4bb0932d8R94

**Comment by avartak** [comment]
@perrotta 

Thanks for the detailed instructions. I will do as you suggest. Just to be sure, I think the RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py is fine. It has all the parameters. 

I realized though, that some parameters were still missing in the fillDescriptions() method. I will add them now and push the changes. I will also change the module label in the fillDescriptions() from "siPixelClusters" to "siPixClusters" so that it's a different string. 

Lastly, I will add a customization to HLTrigger/Configuration/python/customizeHLTforCMSSW.py so as to add the missing parameters to the all the SiPixelClusterProducers

Please correct me in case I miss something. I will push the changes soon.

**Comment by Martin-Grunewald** [comment]
Hi,

If you add the new parameters with values(!) to fillDescriptions, then you do not need an HLT customisation! fillDescriptions will insert the new parameters/values in py configs missing these
parameters for the instances of the plugin in question.
Indeed this is the preferred startegy: use of fillDescriptions.

**Comment by perrotta** [comment]
That's fine, Adish: please go ahead. Anything more descriptive than "siPixClusters" ?

Since you are on it: please also move all the names of the private members of the PixelThresholdClusterizer class to the same style. Up to now you have:
- Most of them start with "the", and bools with "do"
- A few end up with a trailing underscore (please remove)
- Others none of them

The simplest thing would be to make all them start with "the"/"do"; for bools you can use "is" for those required in the checks, e.g. "isBufferAlreadySet"

Moreover:
- "dead_flag" does not need to be a class member, as it is local in https://github.com/cms-sw/cmssw/blob/fdb08e659fd374fc703769bfc585380163ac9fe0/RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc#L468
- Is bufferAlreadySet ever needed/used? Please remove it if not

**Comment by perrotta** [comment]
@Martin-Grunewald indeed...

Adding the missing parameters into fillDescription (the ones that were  
previously accessed with "conf.exists()", I imagine) should be enough,  
and no customization should be needed

Martin Grunewald <notifications@github.com> ha scritto:

**Comment by avartak** [comment]
Hi @Martin-Grunewald , @perrotta 

Adding the parameters to the fillDescription of the PixelThresholdClusterizer does not help. I still see the same error. The error actually, comes from the SiPixelClusterProducer. This class does not have any fillDescription. It simply passes the conf to the PixelThresholdClusterizer here : 

https://github.com/cms-sw/cmssw/blob/fdb08e659fd374fc703769bfc585380163ac9fe0/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.cc#L132

To be sure, do you mean adding fillDescription to the SiPixelClusterProducer class ? 

Adish.

**Comment by Dr15Jones** [comment]
Yes, the `fillDescriptions` function must be added to the actual module.

**Comment by perrotta** [comment]
But "PixelThresholdClusterizer" is an helper class, just one of the possible clusterizers that are used by "SiPixelClusterProducer". Adding the fillDescription() there and remain general enough for any possible clusterizer is an overcomplication, at this point.

I return back to the previous suggestion of mine. You can now add just a customization for the HLT, and when the new configuration from the updated SiPixelClusterizer_cfi.py will be parsed into the HLT ConfDB such a customization will be removed (by TSG).

**Comment by avartak** [comment]
So I made the changes to the variable names, etc. that @perrotta suggested. I also added the customization to HLTrigger/Configuration/python/customizeHLTforCMSSW.py after which I was able to run step2. 

I run into conflicts (as you can see above). May be this is because I am developing in 10_3_0_pre1. Looking into it ....

**Comment by kpedro88** [comment]
@avartak indeed, customizeHLTforCMSSW.py is a frequently modified file. I recommend rebasing to at least the latest prerelease (CMSSW_10_3_0_pre4). There is a guide to rebasing here: http://cms-sw.github.io/tutorial-resolve-conflicts.html

**Comment by avartak** [comment]
Thanks @kpedro88 ... I did the rebasing and now the conflict has gone away

**Comment by kpedro88** [comment]
please test

**Comment by perrotta** [comment]
Ok, thank you @emiglior. It is 
```
Phase2ReadoutMode = cms.int32(-1)
```
in fact...

File: RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc [hunk 5]
```diff
@@ -314,12 +335,38 @@ int PixelThresholdClusterizer::calibrate(int adc, int col, int row)
 {
   int electrons = 0;
 
+  if (doPhase2Calibration) {
+
+    const float gain = theElectronPerADCGain;
+    int p2rm = (thePhase2ReadoutMode < -1 ? -1 : thePhase2ReadoutMode);
```
Comment (perrotta) [inline RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: Some gain in code efficiency could be obtained by replacing L341-358 with
```
    if (thePhase2ReadoutMode <= -1) {
        electrons = int(adc * gain);
    }
    else {
        if (adc < thePhase2KinkADC) {
            electrons = int((adc - 0.5) * gain);
        }
        else {
            const int dualslopeparam = (thePhase2ReadoutMode < 10 ? thePhase2ReadoutMode : 10);
            const int dualslope = int(dualslopeparam <= 1 ? 1. : pow(2, dualslopeparam-1));
            adc -= (thePhase2KinkADC-1);
            adc *= dualslope;
            adc += (thePhase2KinkADC-1);
            electrons = int((adc - 0.5 * dualslope) * gain);
        } 
        electrons += int(thePhase2DigiBaseline);
    }
```

**Comment by avartak** [comment]
@perrotta : I pushed the changes you suggested.

**Comment by kpedro88** [comment]
please test

**Comment by perrotta** [comment]
Thank you Adish!

**Comment by perrotta** [comment]
+1
- Dual slope feature in the phase2 inner tracker digitizer has been implemented in the clusterizer code
- The default Phase2 configuration does not switch it on yet, and therefore there are no changes in the reco outputs

**Comment by Martin-Grunewald** [comment]
+1

**Comment by kpedro88** [comment]
+upgrade

**Comment by fabiocos** [comment]
+1

**Comment by fwyzard** [comment]
Making the new Phase 2 parameters required even if the Phase 2 reconstruction is not used was a mistake.

**Pull request merged**